design/cassandra/design.md

```diff
@@ -0,0 +1,109 @@
+# Kubernetes Operator for Apache Cassandra
+
+## Why ?
+
+* Cassandra is one of the most popular NoSQL databases.
+* Peer-to-Peer write-anywhere design for very high write and read throughput.
+* Despite those advantages, maintaining a Cassandra cluster is a highly manual and tiring task, which requires extensive familiarity with Cassandra's internals.
+* An operator will automate all the manual interventions that a human administrator normally has to perform and provide a production-ready database out-of-the-box.
+
+## Features
+
+The operator will provide all the important features of Cassandra while exposing a user-friendly declarative UI, in accordance to Kubernetes principles.
+More specifically, those features include:
+
+1. Cluster Creation
+2. (Auto)Scaling 
+3. Failed Cassandra Node Replacement
+4. Schedule Node Repair
+5. Backup/Restore
+6. Metrics endpoint for Kubernetes
+
+## Design
+
+### Overview
+
+* This operator will use the pattern: `CRD` + `Controller` + `Sidecar`
+
+![operator-design-overview](media/operator_overview.jpg)
+
+### Cassandra-Kubernetes Mapping
+
+* The design of the Cassandra CRD will follow the Cassandra terminology as much as possible, in order to be familiar to existing users of Cassandra and not confuse new users studying the Cassandra docs.
+
+* Cassandra abstracts resources in the following order:
+`Cluster` -> `Datacenter` -> `Rack` -> `Node`
+
+* We map those abstractions to Kubernetes in the following way:
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
+|     Node      |     Pod       |
+
+An example CRD would look something like this: 
+
+``` yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  version: "0.1"
+  dataCenters:
+  - name: "dc-1"
+    racks:
+    - name: "rack-1"
```

Comment:
I definitely see opportunity to either use, or further refine, the commonality that is captured in the StorageScopeSpec from `rook.io/v1alpha2`: https://github.com/rook/rook/blob/release-0.8/pkg/apis/rook.io/v1alpha2/types.go#L31

The intent of that type is to capture the "boundaries" of the storage substrate, and there is value in using a normalized/common implementation across providers.  Let's see if we can continue this effort of storage resource normalization :)